launchlib/processors.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2016 Palantir Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package launchlib
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+	"io/fs"
+	"math"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	selfCGroup    = "/proc/self/cgroup"
+	selfMountinfo = "/proc/self/mountinfo"
+	cpuGroupName  = "cpu"
+	cpuSharesName = "cpu.shares"
+)
+
+type ProcessorCounter interface {
+	ProcessorCount() (uint, error)
+}
+
+var DefaultCGroupV1ProcessorCounter = CGroupV1ProcessorCounter{
+	fs: os.DirFS("/"),
+}
+
+type CGroupV1ProcessorCounter struct {
+	fs fs.FS
+}
+
+func NewCGroupV1ProcessorCounter(filesystem fs.FS) ProcessorCounter {
+	return CGroupV1ProcessorCounter{fs: filesystem}
+}
+
+func (c CGroupV1ProcessorCounter) ProcessorCount() (uint, error) {
+	cpuCgroupPath, err := c.cpuCGroupPath()
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to get path to cpu cgroup")
+	}
+
+	cpuSharesFilepath := filepath.Join(cpuCgroupPath, cpuSharesName)
+	cpuSharesFile, err := c.fs.Open(c.convertToFSPath(cpuSharesFilepath))
+	if err != nil {
+		return 0, errors.Wrapf(err, "unable to open cpu.shares at expected location: %s", cpuSharesFilepath)
+	}
+	cpuShareBytes, err := io.ReadAll(cpuSharesFile)
+	if err != nil {
+		return 0, errors.Wrapf(err, "unable to read cpu.shares")
+	}
+	cpuShares, err := strconv.Atoi(string(cpuShareBytes))
+	if err != nil {
+		return 0, errors.New("unable to convert cpu.shares value to expected type")
+	}
+	return uint(math.Max(2.0, math.Floor(float64(cpuShares/1024)))), nil
```

Comment:
Possibly worth a link to https://github.com/palantir/go-java-launcher/issues/313 in a comment for context on the 2.0 minimum value.